### DISCUSSION TITLE
cutlass 3.x vs cute: which one to use?

### DISCUSSION THREAD

**Discussion by mgrabban** [NVIDIA/cutlass#1710] [General]
I'm new to cutlass, and currently exploring both cutlass and cute.

The question I have is how is one supposed to use cute? Say I want to write an efficient GEMM for the latest GPU(s)? Should I use cutlass 3.x or cute or both? What would be the preferred way of writing brand new GEMM and related kernels in cutlass?

My understanding is that cutlass 3.x provides higher level [convenience APIs](https://github.com/NVIDIA/cutlass/blob/main/media/docs/gemm_api_3x.md#cutlass-gemm-components) while, based on looking at the cute tutorial examples, it seems, cute might be lower level compared to cutlass 3.x GEMM APIs.

Do I use cutlass 3.x? If yes, then what is the purpose of cute? If cute is the preferred way, can I just ignore cutlass 3.x completely. cute does seem low level. I also notice all cute examples are pure GEMM meaning I do not see any examples that has meaningful epilogues like adding a vector bias, applying an activation (relu, gelu etc.). 

Are there plans to add more cute examples?

Thank you.

**Comment by thakkarV** [comment]
Use whatever floats your boat and gets you the the solution in an expedient manner. There is no one size fits all guideline here.
